### PR TITLE
Use ArrayAccessor and FromIterator in Cast Kernels

### LINKED ISSUE #2137: Use ArrayAccessor and FromIterator in Cast Kernels
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

The cast kernels are currently implemented using macros, and special cased for the various types of array. It would be less code, and easier to maintain if these instead used generics

**Describe the solution you'd like**

Rework the kernel to instead use a combination of `FromIterator` and `ArrayAccessor` instead of macros

**Describe alternatives you've considered**

We could not do this

**Additional context**

#2135

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #2137.

# Rationale for this change
 
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow/src/compute/kernels/cast.rs [hunk 1]
```diff
@@ -270,45 +269,62 @@ pub fn cast(array: &ArrayRef, to_type: &DataType) -> Result<ArrayRef> {
...
+    <T as ArrowPrimitiveType>::Native: ToPrimitive,
+{
+    let mul = 10_f64.powi(scale as i32);
+
+    cast_primitive_to_decimal(
+        array,
+        |v| {
+            // with_precision_and_scale validates the
+            // value is within range for the output precision
+            ((num::ToPrimitive::to_f64(&v).unwrap()) * mul) as i128
```
Comment (tustvold) [inline arrow/src/compute/kernels/cast.rs:323]: I think this is now a checked conversion where previously it wasn't?

File: arrow/src/compute/kernels/cast.rs [hunk 2]
```diff
@@ -270,45 +269,62 @@ pub fn cast(array: &ArrayRef, to_type: &DataType) -> Result<ArrayRef> {
...
+    precision: usize,
+    scale: usize,
+) -> Result<Arc<dyn Array>> {
+    let mul: i128 = 10_i128.pow(scale as u32);
+
+    // with_precision_and_scale validates the
+    // value is within range for the output precision
+    cast_primitive_to_decimal(
+        array,
+        |v| v.to_isize().unwrap() as i128 * mul,
```
Comment (tustvold) [inline arrow/src/compute/kernels/cast.rs:302]: Same here

File: arrow/src/compute/kernels/cast.rs (hunk 1 above)
Comment (viirya) [reply arrow/src/compute/kernels/cast.rs:323]: Hmm, but I cannot use `v as f64` here.

Comment (viirya) [reply arrow/src/compute/kernels/cast.rs:323]:
```
error[E0605]: non-primitive cast: `<T as native::ArrowPrimitiveType>::Native` as `i128`
   --> arrow/src/compute/kernels/cast.rs:300:42
    |
300 |     cast_primitive_to_decimal(array, |v| v as i128 * mul, precision, scale)
    |                                          ^^^^^^^^^ an `as` expression can only be used to convert between primitive types or to coerce to a specific trait object

```

Comment (viirya) [reply arrow/src/compute/kernels/cast.rs]: Added helper functions to `ArrowPrimitiveType`.

File: arrow/src/datatypes/native.rs [hunk 3]
```diff
@@ -79,6 +79,18 @@ pub trait ArrowNativeType:
         None
     }
 
+    /// Convert native type to i128.
+    #[inline]
+    fn to_i128(&self) -> Option<i128> {
```
Comment (tustvold) [inline arrow/src/datatypes/native.rs:84]: I'm not a massive fan of more of these, when I hope to remove them in favour of the standard NUM traits...

File: arrow/src/compute/kernels/cast.rs (hunk 1 above)
Comment (tustvold) [reply arrow/src/compute/kernels/cast.rs]: Why not use AsPrimitive instead of ToPrimitive?

Comment (viirya) [reply arrow/src/compute/kernels/cast.rs]: Ah, good suggestion.

Comment (viirya) [reply arrow/src/compute/kernels/cast.rs]: I'm not aware of `AsPrimitive` trait.

**Review by tustvold — APPROVED** [review]

**Pull request merged**